Perform initial project configuration

aplatkouski opened the pull request:
- [x] ``.nvmrc`` - yнификация используемых версий ``nodejs`` (текущая ``15.2.1``)
- [x] ``package.json``  - перепроверьте, пожалуйста, правильно ли указаны ваши данные в поле ``contributors``.
- [x] ``webpack 5`` - есть нюанс, см. ниже
- [x] ``babel``
- [x] ``.npmrc`` - содержит один параметр ``save-exact=true``, для точного контроля установленных версий пакетов (по зависимостям)

**Унификация стиля написания кода**:
- [x]  ``.editorconfig``
- [x] ``.stylelintrc``
- [x] ``.eslintrc``

**Другие файлы**:
- [x] ``README.md``, ``LICENSE.txt``, ``CONTRIBUTING.md`` and ``CODE_OF_CONDUCT.md``

**Примечание**: В проекте используется 16 версия ``react``, это обсловлено желанием использовать: ``jest``, ``enzyme``, ``enzyme-adapter-react-16`` (последний для 17 версии ещё не выпущен)

## ``.nvmrc``

Почитать про ``nvm`` можно в краткой заметки [Меденджер версий JavaScript](https://guides.hexlet.io/version_managers/).

Кратко:
1. необходимо его установить:
  ```bash
  curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
  ```
2. А после в директории с проектом у кторого настроен ``nvm`` (присутсвует файл ``.nvmrc``) выполнить команду:
  ```bash
  nvm use
  ```

**Замечание**: WebStrom IDE при запуске использует системные настройки (``PATH``) и если в IDE запускать скрипты из ``package.json`` встроенными средствами, то будет использовано ``nodejs`` не из ``.nvmrc``, а системная, и это может привести к ошибке. Решение: запускать WebStrom следующим способом:
1. Перейти в директорию с проектом
2. Выполнить ``nvm use``
3. Запустить WebStrom из текущей директории из командной строки

## ``package.json``

Пользователи WebStorm, могут вызвать на данном файле контекстное меню и выбрать ``Show npm Scripts``, после чего появится отдельный блок ``npm``, из которого удобно запускать скрипты:
  - ``lint`` - выполняет проверку ``*.js`` файлов на соответсвие eslint;
  - ``fix`` - выполняет проверку и исправляет (по возможности) не соответствия
  - ``start`` - запускает приложение (без сборки) в **development** режиме - приложение будет доступно по адресу localhost:8080, при любом изменение файлов входящих в сборку - приложение в браузере будет автоматически обновляться.
  - ``build`` - собирает приложение в **production** режиме и помещает в директорию ``dist``

## ``webpack``

**Замечание**: Из-за того, что в своих проектах я использовал webpack 5, чтобы установить все зависимости нужно использовать доп. ключ --legacy-peer-deps ([подробности](https://github.com/JeffreyWay/laravel-mix/issues/2572)):
```bash
npm install --legacy-peer-deps
````

Ожидаемая структура проекта :
```txt
.
├── src
│   ├── assets
│   │   ├── audio
│   │   │   ├── *.mp3
│   │   │   └── *.wav
│   │   ├── data
│   │   │   └── *.json
│   │   └── images
│   │       ├── *.jpeg
│   │       ├── *.jpg
│   │       └── *.png
│   ├── components
│   │   ├── App.tsx
│   │   └── index.tsx
│   ├── styles
│   │   └── style.scss
│   ├── favicon.ico
│   └── template.ejs
└── public
```

### ``dotenv-webpack``
Для управлениями чувствителными данными (приватные ключи доступа, IP, пароли и т.д.) используется данный плагин.

Для его работы необходимо создать в директории проекта файлы ``.env.development``

```ini
NODE_ENV=development
# your sensitive data
# PRIVATE_KEY=0123456789ABCDEF
```

и файл ``.env.production``

```ini
NODE_ENV=development
# your sensitive data
# PRIVATE_KEY=0123456789ABCDEF
```

### ``webpack.DefinePlugin``

Для получения доступа к инфомации в каком режиме запущено приложение из кода (например для записи отладочной информации) можно использовать:
```js
console.log(process.env.NODE_ENV);
```

где ``NODE_ENV`` содержит одно из двух значений: ``production`` или ``development``


## ``.stylelintrc``

**Замечание**: в WebStrom требуется доп. конфигурация, так как в нём есть встроенный линтер, который может противоречить stylelint (подробности см. в [IntelliJ IDEA: Stylelint](https://www.jetbrains.com/help/idea/using-stylelint-code-quality-tool.html)). Решение: 
  - `Ctrl+Alt+S` -> Languages and Frameworks | Style Sheets | Stylelint -> Enable
  - `Ctrl+Alt+S` -> Editor | Code Style | Style Sheets | CSS -> 
   ```txt
   Tabs and Indents:
     * Tab size: 2
     * Indent: 2
     * Continuation indent: 4
   Other:
      * "HEX Colors"
         - отметить чекбокс "Convert hex colors to" и выбрать "Lower case"
         - отметить чекбокс "Convert hex colors format to" и выбрать "Short format"
    Arrangements:
     * отметить галочку "Sort CSS properties" и выбрать "Custom order" и подставить следующую строку:
    position, top, right, bottom, left, z-index, box-sizing, display, flex, flex-align, flex-basis, flex-direction, flex-wrap, flex-flow, flex-shrink, flex-grow, flex-order, flex-pack, align-content, align-items, align-self, justify-content, order, float, width, min-width, max-width, height, min-height, max-height, padding, padding-top, padding-right, padding-bottom, padding-left, margin, margin-top, margin-right, margin-bottom, margin-left, overflow, overflow-x, overflow-y, -webkit-overflow-scrolling, -ms-overflow-x, -ms-overflow-y, -ms-overflow-style, columns, column-count, column-fill, column-gap, column-rule, column-rule-width, column-rule-style, column-rule-color, column-span, column-width, orphans, widows, clip, clear, font, font-family, font-size, font-style, font-weight, font-variant, font-size-adjust, font-stretch, font-effect, font-emphasize, font-emphasize-position, font-emphasize-style, font-smooth, src, hyphens, line-height, color, text-align, text-align-last, text-emphasis, text-emphasis-color, text-emphasis-style, text-emphasis-position, text-decoration, text-indent, text-justify, text-outline, -ms-text-overflow, text-overflow, text-overflow-ellipsis, text-overflow-mode, text-shadow, text-transform, text-wrap, -webkit-text-size-adjust, -ms-text-size-adjust, letter-spacing, -ms-word-break, word-break, word-spacing, -ms-word-wrap, word-wrap, overflow-wrap, tab-size, white-space, vertical-align, direction, unicode-bidi, list-style, list-style-position, list-style-type, list-style-image, pointer-events, -ms-touch-action, touch-action, cursor, visibility, zoom, table-layout, empty-cells, caption-side, border-spacing, border-collapse, content, quotes, counter-reset, counter-increment, resize, user-select, nav-index, nav-up, nav-right, nav-down, nav-left, background, background-color, background-image, filter, background-repeat, background-attachment, background-position, background-position-x, background-position-y, background-clip, background-origin, background-size, border, border-color, border-style, border-width, border-top, border-top-color, border-top-style, border-top-width, border-right, border-right-color, border-right-style, border-right-width, border-bottom, border-bottom-color, border-bottom-style, border-bottom-width, border-left, border-left-color, border-left-style, border-left-width, border-radius, border-top-left-radius, border-top-right-radius, border-bottom-right-radius, border-bottom-left-radius, border-image, border-image-source, border-image-slice, border-image-width, border-image-outset, border-image-repeat, outline, outline-width, outline-style, outline-color, outline-offset, box-shadow, opacity, -ms-interpolation-mode, page-break-after, page-break-before, page-break-inside, transition, transition-delay, transition-timing-function, transition-duration, transition-property, transform, transform-origin, perspective, appearance, animation, animation-name, animation-duration, animation-play-state, animation-timing-function, animation-delay, animation-iteration-count, animation-direction, animation-fill-mode, fill, stroke
   ```

После выполнения всех вышеперечисленных действий можно привести любой css в WebStrom к требуемому стилю с помощью комбинации ``Ctrl+Shift+Alt+L``

## ``.eslintrc``

**Замечание**: В WebStorm необходимо при первом открытии проекта выполнить доп. настройки:
  - `Ctrl+Alt+S` -> Languages and Frameworks | JavaScript | Code Quality Tools | ESLint - обычно "Automacti ESLint configuration" не плохо справляется, но могут быть ошибки с определением версии node (помжно выбрать "Manual ..." и проверить какую версию он подхватил. По желанию, можно выбрать галочку "Run ESLint on save"

После выполнения всех вышеперечисленных действий можно привести любой js в WebStrom к требуемому стилю с помощью комбинации ``Ctrl+Shift+Alt+L``, и использовать данную команду как альтернативу
```npm run fix```

Очень редко, WebStrom не корректно перености замыкающие квадратные скобки, что противоречит eslint, и тогда без npm run fix не обойтись

aplatkouski committed:
ci: add .nvmrc

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
ci: add .stylelintrc

- add extends stylelint-config-twbs-bootstrap;

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski commented:
Чего в данной конфигурации нет:
 1. Необходимо заменить ``Author`` в ``package.json`` на ``Contributors``. Т.к. поле ``Author`` не позвляет указываь более одного участника
 2. Не учтён TypeScript, React - предлагаю добавить по мере необходимости
 3. конфигурация webpack prod и test разнесена по отдельным файлам, что так же "bad practice"

aplatkouski committed:
ci: add issue and pull request templates

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

kovalenkoiryna15 commented on .editorconfig:
```diff
@@ -0,0 +1,596 @@
+[*]
+charset = utf-8
+end_of_line = lf
+indent_size = 4
```
may be `indent_size = 2` ?

aneekee commented on .eslintrc.json:
```diff
@@ -0,0 +1,16 @@
+    "es2020": true
+  },
+  "extends": [
+    "airbnb-base"
```
Для работы с реактом нужен именно `airbnb`-конфиг, а не `airbnb-base`

aneekee commented on .gitignore:
```diff
@@ -0,0 +1,220 @@
+!.vscode/launch.json
+!.vscode/extensions.json
+*.code-workspace
+
```
Я бы еще добавил `dist` и `package-lock.json`.
`package-lock.json` по-хорошему не нужно игнорить, но если вы это сделаете тут, вы ничего не потеряете + избежите потенциальных git-конфликтов.

aplatkouski replied:
**Done**
```gitignore
# https://github.com/kovalenkoiryna15/rsclone/pull/2#discussion_r551258766
package-lock.json
```

aneekee commented on webpack.config.js:
```diff
@@ -0,0 +1,98 @@
+);
+
+module.exports = {
+  mode: 'development',
```
Можете еще попробовать настроить alias paths, с ними удобнее)

Это чтобы вместо таких импортов:
![image](https://user-images.githubusercontent.com/40535982/103531833-2d30c380-4e9b-11eb-9b70-f302ae24e0a4.png)

были такие:
![image](https://user-images.githubusercontent.com/40535982/103531877-3d48a300-4e9b-11eb-8f44-f979fcfe0865.png)

aplatkouski replied:
К сожалению, я не смог это настроить, т.к. alias прописанные в webpack были проигнорированы eslint

Проблема описана здесь: https://github.com/electron-react-boilerplate/electron-react-boilerplate/issues/1321#issuecomment-349767539

aplatkouski replied:
**Done**

aneekee reviewed (commented):
Я бы добавлял react, typescript, redux сразу, прямо тут.
Без них вы не сможете начать писать код (в моём понимании).
Ну а всякие MaterializeUI или Bootstrap действительно можно добавлять по мере необходимости.

aplatkouski committed:
ci: add .npmrc

- set save-exact=true

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aneekee commented on tsconfig.json:
```diff
@@ -0,0 +1,17 @@
+    "target": "ES5"
+  },
+  "include": [
+    "src/components/index.tsx"
```
В `include` обычно указывают **папки** с исходным кодом.

[Описание поля 'include' в tsconfig](https://www.typescriptlang.org/tsconfig#include)

aplatkouski replied:
fix it
```json
  "include": [
    "src/**/*"
  ]
```

aneekee commented on webpack.config.prod.js:
```diff
@@ -49,31 +54,38 @@ module.exports = {
+        loader: 'awesome-typescript-loader',
+      },
       {
         test: /.css$/,
```
Я бы добавил еще загрузку стилей из scss | sass, но тут уже сами решайте :)

aplatkouski replied:
**Done**
```
      {
        test: /\.s?css$/,
        exclude: /node_modules/,
        use: [MiniCssExtractPlugin.loader, 'css-loader', 'postcss-loader', 'sass-loader'],
      },
```

aplatkouski committed:
ci: add .travis.yml

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
fix: set 'max_line_length = 100' in .editorconfig

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
feat: configure .gitignore

- add 'package-lock.json;
- add '.env.*'.

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
ci: add .babelrc

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
ci: add postcss.config.js

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
feat: replace .eslintrc.js with .eslintrc.json

- configure .eslintrc.json

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
fix: apply eslint and stylelint

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
ci: configure webpack

- webpack.config.js;
- webpack.config.prod.js

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
ci: configure tsconfig.json

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
feat: add and configure 'jest' package

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
feat: add 'alias' to webpack configs and 'paths' to tsconfig.json

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
feat: replace relative paths in imports with alias

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
feat: add LICENSE.txt, CONTRIBUTING.md and CODE_OF_CONDUCT.md files

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
fix: set correct paths in import and remove duplicate import of scss

- App.tsx;

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
ci: add scripts to package.json

- build, fix, lint, start, stats, test, watch;

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski committed:
feat: configure package.json, add template.ejs and styles.scss

- remove 'react-app-env.d.ts';
- remove 'serviceWorker.ts';
- erase field 'scripts' in package.json;
- remove field 'private' from package.json;
- add fields 'contributers', 'main' and others;
- add packages to 'dependencies' and to 'devDependencies'.

Signed-off-by: aplatkouski <5857672+aplatkouski@users.noreply.github.com>

aplatkouski commented:
> Я бы добавлял react, typescript, redux сразу, прямо тут.
> Без них вы не сможете начать писать код (в моём понимании).
> Ну а всякие MaterializeUI или Bootstrap действительно можно добавлять по мере необходимости.

**Done**
